### ISSUE TITLE
Abs() live calculation

### ISSUE THREAD

**Issue by Martin-Rey** [pynbody/tangos#110]
Hi,

I found surprising behaviour when using the abs() live calculation today. On a scalar, negative number, the calculation fails due to an array shape issue:
```
In [3]: halo['mean_iron_hydrogen_ratio']
Out[3]: -2.5548620088265874

In [4]: halo.calculate('abs(mean_iron_hydrogen_ratio)')
---------------------------------------------------------------------------
AxisError                                 Traceback (most recent call last)
<ipython-input-4-64253b3ef2ad> in <module>()
----> 1 halo.calculate('abs(mean_iron_hydrogen_ratio)')

~/Documents/tangos/tangos/core/halo.py in calculate(self, calculation, return_description)
    143         from .. import live_calculation
    144         calculation = live_calculation.parser.parse_property_name_if_required(calculation)
--> 145         (value,), description = calculation.values_sanitized_and_description([self], Session.object_session(self))
    146         if len(value)==1:
    147             retval = value[0]

~/Documents/tangos/tangos/live_calculation/__init__.py in values_sanitized_and_description(self, halos, load_into_session)
    129 
    130         See values_sanitized for the definition of sanitized"""
--> 131         values, desc = self.values_and_description(halos)
    132         return self._sanitize_values(values, load_into_session), desc
    133 

~/Documents/tangos/tangos/live_calculation/__init__.py in values_and_description(self, halos)
    403             input_descriptions.append(id)
    404 
--> 405         calculator, results = self._evaluate_function(halos, input_descriptions, input_values)
    406 
    407         return results, calculator

~/Documents/tangos/tangos/live_calculation/__init__.py in _evaluate_function(self, halos, input_descriptions, input_values)
    520         else:
    521             inherited_description = None
--> 522         return inherited_description, self._as_1xn_array(self._func(halos, *input_values))
    523 
    524 

~/Documents/tangos/tangos/live_calculation/builtin_functions/arithmetic.py in abs(halos, vals)
      7 @BuiltinFunction.register
      8 def abs(halos, vals):
----> 9     return arithmetic_unary_op(vals, functools.partial(np.linalg.norm, axis=-1))
     10 
     11 @BuiltinFunction.register

~/Documents/tangos/tangos/live_calculation/builtin_functions/arithmetic.py in arithmetic_unary_op(vals1, op)
     90         if v1 is not None:
     91             v1 = np.asarray(v1, dtype=float)
---> 92             result = op(v1)
     93         else:
     94             result = None

~/anaconda3/lib/python3.6/site-packages/numpy/linalg/linalg.py in norm(x, ord, axis, keepdims)
   2479             # special case for speedup
   2480             s = (x.conj() * x).real
-> 2481             return sqrt(add.reduce(s, axis=axis, keepdims=keepdims))
   2482         else:
   2483             try:

AxisError: axis -1 is out of bounds for array of dimension 0
```

When used on an actual array, it returns a single number rather than the absolute value of individual elements that I would have expected:
```
In [5]: halo.calculate('abs(dm_density_profile)')
Out[5]: 10607776652.113852
```

I tracked to the use numpy.linalg.norm here:
https://github.com/pynbody/tangos/blob/76ddb305abe547ffb3131284d2ce8e1d4fc7c205/tangos/live_calculation/builtin_functions/arithmetic.py#L7-L9

Is this the expected behaviour and I am misusing the function? Or should the live calculation simply be numpy.abs rather than the full vector norm on the last axis?

Hope this is clear,
Martin

**Comment by Martin-Rey** [comment]
A quick bump on this small issue @apontzen

**Comment by apontzen** [comment]
Sorry for missing this. I think the intended behaviour is vector norm (and we shouldn’t change that intent in case it messes up anyone’s existing code). For example, consider 3D vectors; if `vals` is an `Nx3` array (-> multi-halo) or just a length `3` array,  the implementation is correct.

However I also agree it is confusing that it doesn’t work for getting the absolute value of a single number. I guess we need to do something ugly like call `abs` if `len(vals.shape)==0`. Would this work for your case?

This also makes me wonder what happens in the case of multiple halos, each with a scalar. If memory serves correctly, `vals` in that case should have dimensions `N x 1` and therefore the result will already be OK. But it would be good to be sure; have you tried it?

**Comment by Martin-Rey** [comment]
Ok I understand better the behaviour but it is not as obvious as I thought how to fix it. I have been playing around with several cases for abs applications:
- Abs on a single halo and a scalar ==> vals has shape (1, ) and vals[0] is scalar
- Abs on a single halo and an array ==> vals has shape (1, ) and vals[0] is an array
- Abs on a scalar along the major progenitor branch ==> vals has shape (number_halo_progenitors, ) and vals[0] is a scalar
- Abs on an array along the major progenitor branch ==> vals has shape (number_halo_progenitors, ) and vals[0] is an array
- Abs on multiple halos, each with a scalar ==> vals has shape (number_halo_requested, ) and vals[0] is a scalar
- Abs on multiple halos, each with an array ==> vals has shape (number_halo_requested, ) and vals[0] is an array

The following piece of code testing whether the first element of vals is an array gives satisfactory results (vector norm for arrays and abs on scalars for single and multiple halos) as far as I can tell:
```
    if not hasattr(vals[0], '__len__'):    # Avoid norm failing if abs is called on a single number (issue 110)
        return arithmetic_unary_op(vals, np.abs)
    else:
        return arithmetic_unary_op(vals, functools.partial(np.linalg.norm, axis=-1))
```

If you are happy with it, I will incorporate a test and a fix in a PR.

Martin

**Comment by Martin-Rey** [comment]
I believed this is now fixed by #111
